BLADEBURNER: Typesafety / refactoring

Snarling opened the pull request:
### Non-bladeburner changes
* clampNumber defaults minimum value is -Number.MAX_VALUE instead of Number.MIN_VALUE. MIN_VALUE is almost never something we would want to set as a limit.

### BREAKING CHANGES
* Loading a game from before this PR that was in the middle of Bladeburner will reset your BlackOps progress.
* A savegame from this PR that has an active Bladeburner instance will not load correctly in the main branch, due to lots of static info being removed from the Bladeburner savedata. This has no real impact since the Compression PR was also just merged and causes a similar incompatibility.

### FUTURE TODOS:
* Rework completeAction, completeContract, and completeOperation, implement as Action.complete. Would have dealt with this during this PR, but it was already taking a long time and I wanted to get this out. I'll work on this in the coming weeks.
* 3.0: Get rid of fuzzy matching for player input on enum values.

### Commit details (also just see the commit comments)

458421a: No real substantive changes yet, just jotting down enums that should exist (not all populated yet)
ed8d502: Populated all the created enums with the correct values. They are still not actually used anywhere.
1ef61ef: Made BlackOperations object keyed by enum, and made necessary fixes elsewhere to account for key no longer being a magic string (needs typechecking before access). Couple of other minor changes. See commit details for more info.
b916a51: Remove extra BlackOperations file in data directory, moved descriptions to main BlackOperations.ts file and changed them from React to strings.
23b4e24: Fixed an import
e487134: Changed how black op completion is tracked, see commit details.
94478dd: Fix a test
75833d9: Change format of ActionIdentifier to be typesafe, and other changes to support that. See commit details for more info.
32f8c34: Some changes to GeneralActions and SleeveBladeburnerWork
28178c6: Make Action into an abstract class (GeneralAction is now its own class, so no objects actually use the Action class)
f5085cd: Better directory organization, slim down data saved by Contract/Operation, create a LevelableAction abstract class for Actions that have levels, and move level-related stuff to that class.
18afc2d: Some typesafety stuff for Skills
3838a42: Moving some netscript-specific code out of the Bladeburner class
09ef617: All actions now have a member function for determining if that action is available, use that to streamline some more special-case code
...: Stopped individually writing up every commit. At one point I had Contracts and Operations acting as static objects just like GeneralActions and BlackOperations, but I reverted that due to some issues around save/load.
...

Snarling committed:
Restart Bladeburner typesafety PR

Skeleton of Enums.ts
Rename BlackOperationName to BladeBlackOpName (consistency with other Bladeburner enums)

Snarling committed:
Populate enums with values

Snarling committed:
BlackOp changes

BlackOperation.tsx -> .ts
BlackOperations object is keyed by the correct enum instead of string
Populated during declaration instead of immediately after using an IIFE
Blackop objects is already not part of the save file -> remove reviver functionality and force constructor to take params
Remove resetBlackOps function which was being used to keep the static BlackOperations object in sync with the loaded save. Now just update on game load and when creating a new Bladeburner instance.
Some fixes to properly typecheck BlackOpName usage

Snarling committed:
Move blackop descript into actual object

Move BlackOp descriptions to BlackOperations.ts.
Verified all bladeburner descriptions are plaintext -> changed from React elements to strings
Remove BlackOperations.tsx file in data directory - this only had descriptions previously.
Minor changes in BlackOpElem.tsx

Snarling committed:
Fix an import

Minor change to an import statement

Snarling committed:
Remove bladeburner.blackops property

* Add numBlackOpsCompleted property
* Add savegame conversion to new property
* Remove blackops property that was being used to track which had been completed
* Add a static array containing the sorted list of blackOps
* Some slight refactoring to accomodate new way of tracking blackop completion

Snarling committed:
Update test snapshot for previous commit

Snarling committed:
ActionIdentifier is typesafe

* Get rid of weird numerical type
* General is a type
* ActionIdentifier is a simple object rather than a class. Added a loading helper for converting the old format (with numerical types) to the new type.
* Idle is not an identifier, it is null
* Operation name is typechecked
* BlackOperation no longer extends Operation (necessary because of conflicting type for name property)
* Still need to do testing for automation changes, to make sure no change in behavior there

Snarling committed:
GeneralActions

* Consolidated the two existing GeneralActions.tsx files into a new GeneralActions.ts file
* Added a new class for GeneralAction
* Some changes at SleeveBladeburnerWork to make that typesafe, and make it load into idle if the loaded data is not valid. Did not add specific loading code for version change, so any sleeves performing bladeburner work are expected to load in as idle on first load.

Snarling committed:
Action is an abstract class

Nothing is initialized as an Action, and nothing is saved to the save file as an Action (GeneralAction previously used the Action class directly, but those were not part of the save file either).

Snarling committed:
A lot of stuff

* Reorganized files: Classes for action types are in the Actions folder
* Static objects containing the actions are in the Data folder, added static objects for Contract and Operation
* Contracts and Operations no longer save/load a bunch of static data. The static data is initialized once on-demand and reused, and only dynamic data is saved/loaded.
* Removed some properties from the generic Action class and moved them to a new LevelableAction.
* No more separate growths file, now stored on the levelable action
* Probably some other minor stuff

Snarling committed:
Skills

* Typesafety for accessing skills by name. Skill multipliers are not yet typesafe.
* Skills have a default max level of MAX_SAFE_INTEGER
* Added a new clamping function specific to integer values, applied it to skill levels
* Streamlined some netscript functions

Snarling committed:
Start of netscript changes

* Start moving the netscript-specific functionality over to the NetscriptFunctions file
* Rename bladeburner.upgradeSkill to increaseSkill, makes it more clear that it doesn't handle cost of upgrading
* Move some more level-related properties off the top level ActionClass and down to LevelableAction
* Formatting change for "BlackOpsAttempt"

Snarling committed:
More streamlining

* Remove reqdRank property from Operations, it was not used
* Added id field to all action types to quickly get a valid ActionIdentifier from an Action.
* Renamed sequencing variable for BlackOp from "id" to "n"
* Deduplicate some more functionality between API and UI
* Added getAvailability function to all action types, allows getting rid of lots of "special case" code sections
* Renamed bladeburner.create function to reset, it's not a constructor it's just resetting some static data

Snarling committed:
Reorganization

* Create a Types file for bladeburner for common non-class/non-enum types
* Rewrite SuccessChance component
* Some import cleanups on affected files

Snarling committed:
Slimming Bladeburner.ts

* Move getActionIdFromTypeAndName to ActionIdentifier.ts
* Remove parseCommandArguments, reuse parseCommand function from terminal parsing.

Snarling committed:
Move some more stuff

* Renamed ActionIdentifier.ts to Actions/utils.ts
* Moved some more Action utility functions that don't need access to bladeburner over to utils.ts
* Change getActionIdFromTypeAndName to getActionFromTypeAndName - it was always immediately being followed up by getting the action from the actionId so the actionId was just a pointless middle step
* Startbutton now has the action instead of the actionId

Snarling committed:
Fix import issues

* Moved assertLoadingType into its own file to prevent import collision
* Moved loadActionIdentifier into its own file to prevent import collision

Snarling committed:
Update safe loading method

Make the loading method work a bit better as far as ensuring bladeburner.contracts and bladeburner.operations are equal to the static versions and have no discrepancies regardless of possible savedata format differences. Also the loaded properties are all typechecked.
Also removed the `name` property from savedata on these objects, as it was no longer needed.

Snarling committed:
More stuff

* Improved "Availability" type, used for upgrading skill cost.
* Removing unused code or moving to more suitable location
* Remove recruitment-specific functions in bladeburner.ts, moved this so it's defined in the recruitment action itself

Snarling committed:
Merge branch 'dev' into bladeburnerTypesafety

Snarling committed:
Operations and Contracts as getters

Actual static objects are local and not exported, only export a getter which also removes the need for externally calling the initializer.

Snarling committed:
Revert "Operations and Contracts as getters"

Revert getter commit, it causes the initializer to run right away (due to the getter being called during export), which causes import issues.

Snarling commented:
~~There's an unfortunate issue with the design I went with on Contracts and Operations, where these classes hold all of the information about the object, and the objects are statically initialized with the dynamic content loaded in when loading a bladeburner object.~~

~~The Bladeburner object can be loaded in instances where we're not actually trying to load the game yet, e.g. when doing the import comparison thing. This would trigger the code that loads the contracts+operations savedata onto the static objects even if the player doesn't actually go through with the import.~~

~~Might have to switch over to something more like how Factions works - where the actual object that interacts with savedata is minimal, and there's a separate ContractsData or something that has all of the static data. I don't really like that idea though as it would require a lot of changes to Actions in general, makes it harder to interact with the data, and just generally is kind of an ugly solution. I will think a little more on a better way to handle this.~~
Addressed in 453ddaf, it was a lot simpler than I thought it would be,

Snarling committed:
Contracts and Operations no longer static

Static contracts and operations had issues with import comparison (data would load onto static objects even though the save wasn't actually being loaded). They are no longer static, but still typesafe with the same loading validation.
This also forced me to move gertActionObject and getActionFromTypeAndName back to the Bladeburner class.

Snarling committed:
Fix an invalid DOM nesting console error

d0sboots commented:
> Addressed in https://github.com/bitburner-official/bitburner-src/commit/453ddaf3ceaa88fd5fd6d6bac972b33ec658b688, it was a lot simpler than I thought it would be

Nice, because that's the design I'd prefer as well :D

Snarling committed:
Attempt type

* Added an Attempt type (basically a mirror of Availability, but for actually trying to perform a task instead of just checking if it can be performed).
* startAction returns an Attempt.
* Added an internal joinFaction function which returns an Attempt
* Get rid of last netscript specific code in bladeburner class
* Simplified netscript startAction and joinBladeburnerFaction code

Snarling commented:
Test failure is just due to removing this.reset() from Bladeburner constructor, so contracts/operations are only randomized in count once instead of twice, and end up with different counts than the snapshots. I'll include a test fix in the next commit.

Snarling committed:
More skill stuff

* Change keys of weights and decays on action to same keys as player skills
* Add another enum for skill multipliers
* Some more minor tweaks
* TODO: skill multipliers shouldn't be saved, as it's entirely calculated data

Snarling committed:
Remove skillMultipliers from save

* skillMultipliers is no longer saved
* Due to use of getKeyList, constructor needs to be able to be called right away., but it was dependent on Player. Moved the dependent code into an `init` function.

Snarling committed:
City

* Rewrite some code in City
* Also fixed a mistake in clampNumber where Number.MIN_VALUE is used as the default minimum clamp which is obviously wrong (we almost never want something to default to that value). Didn't affect any existing use cases as they all had explicitly defined minimums. Default should either be 0 or -Number.MAX_VALUE, I went with the latter because it's more cumbersome to specify in cases where that should be the default, it's easy to specify 0.
* Also changed clampInteger to have a similar default, and changed all the existing use cases to specify 0 if they need it.

Snarling committed:
Fix contracts count bug

Snarling committed:
Remove non-working savegame conversion

This savegame conversion code doesn't work anyway, due to the use of getKeyList the load process will not load in the value for the old "blackops" property.
The only impact here is that savegames loaded into the new version while in the middle of bladeburner will have to restart the blackops from the beginning, which should only be a minor setback.

Snarling committed:
Some final changes before merging

* Move loadActionIdentifier to a utils folder
* clampInteger uses Math.round instead of Math.floor
* Some minor formatting changes

Snarling committed:
Fix default values for stamina

This was causing clampNumber to encounter a NaN.
